ZKUI-161: Listing Policy

AsmaaELM opened the pull request:
Issue: https://scality.atlassian.net/browse/ZKUI-161

bert-e commented:
# Hello asmaaelm,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/zenko-ui/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.

ChengYanJin commented on src/react/queries.ts:
```diff
@@ -83,4 +83,12 @@ export const getUserListUsersQuery = (accountName: string, IAMClient: IAMClient)
+});
+
+export const getUserPoliciesQuery = (UserName: string, IAMClient: IAMClient) => ({
+  queryKey: ['listIAMUserPolicies', UserName],
```
```suggestion
  queryKey: ['listPolicies', UserName],
```

ChengYanJin commented on src/react/queries.ts:
```diff
@@ -83,4 +83,12 @@ export const getUserListUsersQuery = (accountName: string, IAMClient: IAMClient)
   refetchOnWindowFocus: false,
+});
+
+export const getUserPoliciesQuery = (UserName: string, IAMClient: IAMClient) => ({
```
The name is bit misleading. I would suggest to rename
```suggestion
export const getPoliciesQuery = (UserName: string, IAMClient: IAMClient) => ({
```

ChengYanJin commented on src/js/IAMClient.ts:
```diff
@@ -154,6 +154,15 @@ export default class IAMClient implements IAMClientInterface {
       .promise();
   }
 
+  deleteUserPolicy(userName: string, policyName: string) {
+    return notFalsyTypeGuard(this.client)
+      .deleteUserPolicy({
+        UserName: userName,
+        PolicyName: policyName
+      })
+      .promise();
+  }
+
```
the endpoint to should use is delete-policy not delete-user-policy

ChengYanJin commented on src/react/account/AccountContent.tsx:
```diff
@@ -24,6 +24,18 @@ function AccountContent() {
         <Route path={`${path}/users/:IAMUserName/update-user`}>
           <AccountUpdateUser />
         </Route>
+        <Route path={`${path}/create-user-policy`}>
+          {/*TODO: when the subticket 'create a new policy' is ready*/}
+          {/* <AccountCreateUserPolicy /> */}
+        </Route>
+        <Route path={`${path}/attach-user-policy`}>
+          {/*TODO: when the Attachment subtask ZKUI-163 is ready*/}
+          {/*<AccountAttachUserPolicy />*/}
+        </Route>
+        <Route path={`${path}/users/:IAMUserName/update-user-policy`}>
+          {/*TODO: when the Attachment subtask ZKUI-163 is ready*/}
+          {/*<AccountUpdateUserPolicy /> */}
+        </Route>
```
We likely don't want to expose those route until the subtasks are done.

JBWatenbergScality commented on src/react/account/AccountUserPoliciesList.tsx:
```diff
@@ -0,0 +1,306 @@
+              isLoading={listPoliciesQuery.status === 'loading'}
+              tooltipOverlay="Search is disabled while loading policies"
+            >
+              <SearchInputComponent
```
Policies counting is missing

ChengYanJin commented on src/react/account/AccountUserPoliciesList.tsx:
```diff
@@ -0,0 +1,306 @@
+
+const RenderEditButton = ({ policyName }: { policyName: string }) => {
+  const history = useHistory();
+  const disableCondition = ['storage-account-owner-policy', 'storage-manager-policy'].includes(policyName);
```
I would suggest defining a constant for 'storage-account-owner-policy', 'storage-manager-policy'

JBWatenbergScality reviewed (commented):
To me we should try to figure out if we can build a component factorizing a bit the logic with the users listing as I mentioned before. Because most of the table logic is the same, same erro management. same counting of entities, same search behavior.

JBWatenbergScality commented on src/react/account/__tests__/AccountUserPoliciesList.test.tsx:
```diff
@@ -0,0 +1,145 @@
+  rest.post(`${TEST_API_BASE_URL}/`, (req, res, ctx) => {
+    return res(
+      ctx.xml(`
+    <ListUsersResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">
+        <ListUsersResult>
+            <Users>
+                <member>
+                    <UserId>${SAMPLE_USER_ID}</UserId>
+                    <Path>${SAMPLE_USER_PATH}</Path>
+                    <PolicyName>${SAMPLE_Policy_NAME}</PolicyName>
+                    <Arn>${SAMPLE_ARN}</Arn>
+                    <ModifiedDate>${SAMPLE_CREATE_DATE}</ModifiedDate>
+                </member>
+            </Users>
+            <IsTruncated>false</IsTruncated>
+        </ListUsersResult>
+        <ResponseMetadata>
+            <RequestId>61221a552b4592e5b784</RequestId>
+        </ResponseMetadata>
+    </ListUsersResponse>
```
We should not mock ListUsersResponse here but ListPoliciesResponse I think

JBWatenbergScality commented on src/react/account/AccountUserPoliciesList.tsx:
```diff
@@ -0,0 +1,306 @@
+
+const SEARCH_QUERY_PARAM = 'search';
+
+const AccountUserPoliciesList = ({ accountName }: { accountName?: string }) => {
```
```suggestion
const AccountPoliciesList = ({ accountName }: { accountName?: string }) => {
```

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+const renderActionButtons = (rowValues, accountName) => {
+  const { arn, policyName } = rowValues;
+  return (
+    <div style={{ display: 'flex' }}>
```
there is a box component from core-ui 
```suggestion
    <Box display='flex' }}>
```

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+          return (
+            <>
+              <Tooltip
+                overlay={'This is a Preseeded (predefined?) Scality Policy'}
```
```suggestion
                overlay={'This is a predefined Scality Policy'}
```
And We are missing a condition here. We only display this tooltip for Scality Internal Policies.

ChengYanJin commented on src/react/account/AwsPaginatedResourceTable.tsx:
```diff
@@ -0,0 +1,121 @@
+              display: 'flex',
+              alignItems: 'center',
+            }}
+          >
```
Please use Box component from core-ui

ChengYanJin commented on src/react/account/AccountUserList.tsx:
```diff
@@ -23,6 +19,13 @@ import {
   getUserListGroupsQuery,
 } from '../queries';
+
+import user from '../reducers/user';
```
could be removed

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+          setShowModal(true);
+        }}
+        variant='danger'
+        tooltip={{ overlay: listPoliciesStatus === 'loading' ? 'Deleting...':'You cannot delete a predefined Scality Policy', placement: 'right' }}
```
Not sure why we want to place the tooltip on the right? since we place the actions on the right, I am afraid we won't have any space.

ChengYanJin replied:
We are missing a condition here. We only display this tooltip for scality internal policies.

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+const RenderEditButton = ({ policyName }: { policyName: string }) => {
+  const history = useHistory();
+  const storageAccountOwnerPolicy = 'storage-account-owner-policy';
+  const storageManagerPolicy = 'storage-manager-policy';
```
I would put it outside the RenderEditButton component. because likely they will be used in other places.
I would create a group 
```js
const SCALITY_INTERNAL_POLICIES= [STORAGE_ACCOUNT_OWNER_POLICY, STORAGE_MANAGER_POLICY, DATA_CONSUMER_POLICY]```

ChengYanJin replied:
```suggestion
  const STORAGE_MANAGER_POLICY = 'storage-manager-policy';
```

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+
+const RenderEditButton = ({ policyName }: { policyName: string }) => {
+  const history = useHistory();
+  const storageAccountOwnerPolicy = 'storage-account-owner-policy';
```
```suggestion
  const STORAGE_ACCOUNT_OWNER_POLICY = 'storage-account-owner-policy';
```

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+        icon={<i className='fa fa-pen'></i>}
+        onClick={() => history.push(`users/${policyName}/update-policy`)}
+        tooltip={{ overlay: disableCondition ? 'You cannot edit this predefined Scality Policy' : '', placement: 'right' }}
+      />
```
In theory, this should not be a part of this PR since the feature (Create Policy / Delete policy / Edit policy) is not ready.

But since the code is already here, what I would suggest is to put the policy feature under the feature gate same as we did for IAM user.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.4.7`

* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+        label='Edit'
+        icon={<i className='fa fa-pen'></i>}
+        onClick={() => history.push(`users/${policyName}/update-policy`)}
+        tooltip={{ overlay: disableCondition ? 'You cannot edit this predefined Scality Policy' : '', placement: 'right' }}
```
We are missing a condition here. We only display this tooltip for scality internal policies.

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+      <Button
+        disabled
+        icon={<i className='fas fa-trash' />}
+        style={{ height: spacing.sp24, marginLeft: '0.6rem' }}
```
We should really avoid this inline styling. please wrap with a Box component

ChengYanJin commented on src/react/account/AwsPaginatedResourceTable.tsx:
```diff
@@ -0,0 +1,121 @@
+  </>
+);
+
+const AwsPaginatedResourceTable = ({additionalHeaders , columns, query: {getResourceQuery, getEntitiesFromResult, prepareData}, defaultSortingKey, getItemKey, labels: {disabledSearchWhileLoading, pluralResourceName, singularResourceName, loading, errorPreviousHeaders, errorInTableContent }} : Props) => {
```
Please add typing for the props

ChengYanJin commented on src/react/account/AwsPaginatedResourceTable.tsx:
```diff
@@ -0,0 +1,121 @@
+  const data = prepareData(queryResult, search);
+
+  return (
+    <div
+      style={{
+        height: '100%',
+      }}
+    >
```
Box component

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,234 @@
+        };
+      });
+
+      if (search) {
+        return iamPolicies.filter((user) =>
+          user.policyName?.toLowerCase().startsWith(search.toLowerCase()),
+        );
+      }
```
Not sure we want to search only by policy name. To be discussed with @Cuervino

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.4.7`

* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.

AsmaaELM committed:
Listing policy

ChengYanJin commented on src/js/utils.ts:
```diff
@@ -1,3 +1,7 @@
+const STORAGE_ACCOUNT_OWNER_POLICY = 'storage-account-owner-policy';
+const STORAGE_MANAGER_POLICY = 'storage-manager-policy';
+const DATA_CONSUMER_POLICY = 'data-consumer-policy';
+export const SCALITY_INTERNAL_POLICIES = [STORAGE_ACCOUNT_OWNER_POLICY, STORAGE_MANAGER_POLICY, DATA_CONSUMER_POLICY];
```
I am not sure you are really using those constants here.

AsmaaELM replied:
Yeah, but I thought that we wanted to centralize those constants in the utils file for other need.
I'll remove it then!

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,244 @@
+
+  const prepareData = (queryResult, search) => {
+    if (queryResult.firstPageStatus === 'success') {
+      const iamPolicies = queryResult && queryResult.data && queryResult.data.map((user) => {
```
```suggestion
      const iamPolicies = queryResult && queryResult.data && queryResult.data.map((policy) => {
```

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,244 @@
+        return iamPolicies.filter(user =>
+          Object.values(user).find(val => val?.toString().toLowerCase().startsWith(search) )
+        );
+      }
```
as discussed with Valamir, I think we shouldn't need this filter anymore.

AsmaaELM replied:
Yeah, sure. But when I remove this part (or put it on a comment) it doesn't resolve the problem.
Because we cannot search globally as discussed with Valamir.
So do you want me to remove it anyway?

ChengYanJin replied:
No. We will need the global search for this list policy table. I pushed a commit in order to achieve this.

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,244 @@
+      const iamPolicies = queryResult && queryResult.data && queryResult.data.map((user) => {
+        return {
+          policyPath: user?.Path.substring(1),
+          policyName: user?.PolicyName + '_' + user?.Path,
```
I am wondering wouldn't this be enough?
```suggestion
          policyName: user?.PolicyName,
```

AsmaaELM replied:
Yeah, sorry. I add it in order to address your previous comment: "We are missing a condition here. We only display this tooltip for Scality Internal Policies".
Do you have please any other solution in order to pass the path that we need in the condition (in order to enable/disable the tooltip) near the policyName ?

ChengYanJin commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,244 @@
+        minWidth: '18rem',
+      },
+      Cell({ value: policyParams }: { value: string }) {
+          const [policyName, policyPath] = policyParams.split('_');
```
We can get the policy path directly from `listPolicies` API, I don't understand why we need to aggregate `policyName` and `policyPath` and split them..

AsmaaELM replied:
Same here, we need the policyPath for enabling the tooltip near the policyName

ChengYanJin replied:
In this `policyName` Cell, we can get the rowValues which includes the `policyPath`.

ChengYanJin commented on src/react/account/AccountUserList.tsx:
```diff
@@ -104,49 +118,90 @@ const renderActionButtons = (rowValues) => {
+        cancel={() => setShowModal(false)}
+        approve={() => {
+          deleteUserMutation.mutate(userName);
+        }}
+        titleText={`Permanently remove the following user ${userName} ?`}
+      />
+
+      <Button
+        id="delete-accessKey-btn"
+        disabled={
+          (accessKeysResult && accessKeysResult?.length >= 1) ||
+          (listGroupsResult && listGroupsResult.Groups?.length >= 1) ||
+          accessKeyStatus === 'loading' ||
+          listGroupStatus === 'loading'
+        }
+        icon={<i className="fas fa-trash" />}
+        style={{ height: spacing.sp24, marginLeft: '0.6rem' }}
+        label="Delete"
+        onClick={() => {
+          setShowModal(true);
+        }}
+        variant="danger"
+        tooltip={{
+          overlay:
+            accessKeyStatus === 'loading' ? 'loading...' : 'Remove accessKey',
+          placement: 'right',
+        }}
+      />
+      {accessKeyStatus === 'error' && (
+        <Banner
+          icon={<i className="fas fa-exclamation-triangle" />}
+          title="Error: Unable to delete user"
+          variant="danger"
+        >
+          Error: Unable to delete user.
+        </Banner>
+      )}
+    </>
+  );
+};
```
I believe the feature of Delete User is backportted from development/1.5. 
In term of good practice, I would put them in a dedicated commit.

AsmaaELM replied:
I added a new commit for this part

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.4.8`

* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.

AsmaaELM committed:
Delete user

ChengYanJin committed:
ZKUI-161: Add filterData logic to manage the search

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.4.9`

* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.

JBWatenbergScality commented on src/react/account/AccountDetails.tsx:
```diff
@@ -52,6 +53,9 @@ function AccountDetails({ account }: Props) {
       {features.includes("IAM_USERS") && <CustomTabs.Tab label="Users" path={`${url}/users`}>
         <AccountUserList accountName={accountName} />
       </CustomTabs.Tab>}
+      {features.includes("IAM_USERS") && <CustomTabs.Tab label="Policies" path={`${url}/policies`}>
+        <AccountPoliciesList accountName={accountName} />
+      </CustomTabs.Tab>}
```
nit
```suggestion
      {features.includes("IAM_USERS") && <><CustomTabs.Tab label="Users" path={`${url}/users`}>
        <AccountUserList accountName={accountName} />
      </CustomTabs.Tab>
      <CustomTabs.Tab label="Policies" path={`${url}/policies`}>
        <AccountPoliciesList accountName={accountName} />
      </CustomTabs.Tab></>}
```

AsmaaELM replied:
Thanks for your suggestion, but the tab doesn't display when we put it inside a tag like <> or <div> ...

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+        variant="secondary"
+        label="Edit"
+        icon={<i className="fa fa-pen"></i>}
+        onClick={() => history.push(`users/${policyName}/update-policy`)}
```
```suggestion
        onClick={() => history.push(`/policies/${policyName}/update`)}
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+const STORAGE_ACCOUNT_OWNER_POLICY = 'storage-account-owner-policy';
+const STORAGE_MANAGER_POLICY = 'storage-manager-policy';
+
+const RenderEditButton = ({
```
```suggestion
const EditButton = ({
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  path: string;
+}) => {
+  const history = useHistory();
+  const isEditPolicyDisabled = [
+    STORAGE_MANAGER_POLICY,
+    STORAGE_ACCOUNT_OWNER_POLICY,
+  ].includes(policyName);
```
We should likely check the policy name and path to avoid false positive where a customer name it's custom policy `STORAGE_MANAGER_POLICY ` or `STORAGE_ACCOUNT_OWNER_POLICY `

AsmaaELM replied:
Sorry, I'm not sure to understand. To be discussed ...

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  );
+};
+
+const RenderAttachButton = () => {
```
```suggestion
const AttachButton = () => {
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+        variant="secondary"
+        label="Attach"
+        icon={<i className="fas fa-link"></i>}
+        onClick={() => history.push('attach-user-policy')}
```
I'm not convinced about this path. Likely we could change it to:
```suggestion
        onClick={() => history.push(`/policies/${policyName}/attachments`)}
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  );
+};
+
+const renderActionButtons = (rowValues, accountName) => {
```
```suggestion
const ActionButtons = ({rowValues, accountName}: TYPE_TODO) => {
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  const { policyName, path, arn, accountName } = rowValue;
+  const IAMClient = useIAMClient();
+  const [showModal, setShowModal] = useState(false);
+  const pathCondition = path.includes('scality-internal');
```
```suggestion
  const isInternalPolicy = path.includes('scality-internal');
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+      <Box ml="0.6rem">
+        <Button
+          style={{ height: spacing.sp24 }}
+          disabled={pathCondition}
```
```suggestion
          disabled={isInternalPolicy}
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+          variant="danger"
+          tooltip={{
+            placement: 'top',
+            overlay: pathCondition
```
```suggestion
            overlay: isInternalPolicy
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  const { arn, policyName, policyPath } = rowValues;
+  return (
+    <Box display="flex">
+      <RenderAttachButton />
```
```suggestion
      <AttachButton />
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  return (
+    <Box display="flex">
+      <RenderAttachButton />
+      <RenderEditButton policyName={policyName} />
```
```suggestion
      <EditButton policyName={policyName} />
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  const IAMClient = useIAMClient();
+  const [showModal, setShowModal] = useState(false);
+  const pathCondition = path.includes('scality-internal');
+  const { status: listPoliciesStatus } = useQuery(
+    getPoliciesQuery(policyName, notFalsyTypeGuard(IAMClient)),
+  );
```
`getPoliciesQuery(policyName,` is a bit confusing, is it listing several policies or just one as the signature expecting a policyName as a first parameter makes me feel ? or maybe is it expecting the `accountName` instead of the `policyName` 🤔  ?

JBWatenbergScality replied:
I'm not sure about the value of listing policies in this delete screen

AsmaaELM replied:
This part helps us to get the status that we use for handling errors. Do you want me to remove this part?

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+          }}
+        />
+      </Box>
+      {listPoliciesStatus === 'error' && (
+        <Banner
+          icon={<i className="fas fa-exclamation-triangle" />}
+          title="Error: Unable to delete policy"
+          variant="danger"
+        >
+          Error: Unable to delete policy.
+        </Banner>
+      )}
```
The error text and the condition are not matching. This message should be displayed if the actual deletion failed, not if we failed to list something

ChengYanJin replied:
After discussing with @Cuervino , we decided to handle the error as how it is today, and a workshop to handle errors more global will come.

AsmaaELM replied:
Does this mean that I can keep for now this part like or I should change it?

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+
+const AccessPolicyNameCell = (rowValue) => {
+  const { policyPath, policyName } = rowValue;
+  const enableTooltip = policyPath.includes('scality-internal');
```
```suggestion
  const isInternalPolicy = policyPath.includes('scality-internal');
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+    getPoliciesQuery(notFalsyTypeGuard(accountName), IAMClient);
+  const getEntitiesFromResult = (data) => data.Policies;
+
+  const prepareData = (queryResult) => {
```
would be nice to type this function

AsmaaELM replied:
Sorry, I didn't find the convenient type for this function. Do you have any suggestions?

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+  );
+};
+
+const AccessPolicyNameCell = (rowValue) => {
```
```suggestion
const AccessPolicyNameCell = ({rowValue}: TYPE_TODO) => {
```

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+      cellStyle: {
+        minWidth: '18rem',
+      },
+      Cell: (value) => AccessPolicyNameCell(value.row.original),
```
```suggestion
      Cell: (value) => <AccessPolicyNameCell rowValue={value.row.original} />,
```

AsmaaELM replied:
Using this syntax I'm getting errors. To be discussed ...

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,281 @@
+        minWidth: '5rem',
+      },
+      disableSortBy: true,
+      Cell: (value) => renderActionButtons(value.row.original, accountName),
```
```suggestion
      Cell: (value) => <ActionButtons rowValue={value.row.original} accountName={accountName} />,
```

AsmaaELM replied:
Same here ...

JBWatenbergScality commented on src/react/account/AwsPaginatedResourceTable.tsx:
```diff
@@ -0,0 +1,174 @@
+  useAwsPaginatedEntities,
+} from '../utils/IAMhooks';
+import { TitleRow as TableHeader } from '../ui-elements/TableKeyValue';
+import React, { ChangeEvent, useMemo } from 'react';
```
useMemo import is unused

JBWatenbergScality commented on src/react/account/AwsPaginatedResourceTable.tsx:
```diff
@@ -0,0 +1,174 @@
+                  }}
+                />
+              ) : (
+                <Table.SearchWithQueryParams queryParams={SEARCH_QUERY_PARAM} />
```
I think we are missing the disabled state here :(

AsmaaELM replied:
To be discussed with @ChengYanJin

JBWatenbergScality commented on src/react/account/__tests__/AccountPoliciesList.test.tsx:
```diff
@@ -0,0 +1,140 @@
+      expect(deleteButton).toBeDisabled();
+
+
+  });
```
We should test the deletion of a non internal policy as well

JBWatenbergScality commented on src/react/queries.ts:
```diff
@@ -83,4 +81,13 @@ export const getUserListUsersQuery = (accountName: string, IAMClient: IAMClient)
+  enabled: IAMClient !== null,
+  onError: console.log,
+  refetchOnMount: false,
+  refetchOnWindowFocus: false
```
```suggestion
export const getPoliciesQuery = (accountName: string, IAMClient: IAMClient) => ({
  queryKey: ['listPolicies', accountName],
  queryFn: (maxItems: number, marker: string) => IAMClient.listPolicies(1000, marker),
  enabled: IAMClient !== null,
  refetchOnMount: false,
  refetchOnWindowFocus: false
```

ChengYanJin commented on src/react/account/AccountUserList.tsx:
```diff
@@ -118,10 +118,81 @@ const renderActionButtons = (rowValues) => {
+        variant="danger"
+        tooltip={{
+          overlay:
+            accessKeyStatus === 'loading' ? 'loading...' : 'Remove accessKey',
```
I believe this is not the correct tooltip message.

ChengYanJin commented on src/react/account/AccountUserList.tsx:
```diff
@@ -118,10 +118,81 @@ const renderActionButtons = (rowValues) => {
+          placement: 'right',
+        }}
+      />
+      {accessKeyStatus === 'error' && (
+        <Banner
+          icon={<i className="fas fa-exclamation-triangle" />}
+          title="Error: Unable to delete user"
+          variant="danger"
+        >
+          Error: Unable to delete user.
+        </Banner>
+      )}
```
To be removed and add model to display error.

ChengYanJin commented on src/react/account/AccountUserList.tsx:
```diff
@@ -118,10 +118,81 @@ const renderActionButtons = (rowValues) => {
+
+      <Button
+        id="delete-accessKey-btn"
+        disabled={
+          (accessKeysResult && accessKeysResult?.length >= 1) ||
+          (listGroupsResult && listGroupsResult.Groups?.length >= 1) ||
+          accessKeyStatus === 'loading' ||
+          listGroupStatus === 'loading'
+        }
```
We should also disable the Delete button when listUser is in loading status.

ChengYanJin committed:
Fix Deletion of User

ChengYanJin committed:
Fix the dependency issue

ChengYanJin committed:
Configure npm with legacy peers

NOTE: If you create your package-lock.json file by running npm install
with flags that can affect the shape of your dependency tree, such as
--legacy-peer-deps or --install-links, you must provide the same flags
to npm ci or you are likely to encounter errors. An easy way to do this
is to run, for example, npm config set legacy-peer-deps=true
--location=project and commit the .npmrc file to your repo.

JBWatenbergScality commented on src/react/account/AccountPoliciesList.tsx:
```diff
@@ -0,0 +1,331 @@
+  return (
+    <>
+      {isInternalPolicy && (
+        <Tooltip
+          overlay={'This is a predefined Scality Policy'}
+          overlayStyle={{
+            width: '13rem',
+          }}
+        >
+          {policyName} <Icon className="fas fa-question-circle fa-xs"></Icon>
+        </Tooltip>
```
nit - can wait for later - can be ignored for now.

I wonder wether we already have(in zenko-ui) or should have such Tooltip+questionCircle icon component in core-ui. Something like `<HelperText overlay={string}>{icon => <>Toto {icon}</> || 'toto'}</HelperText>`
if the children props is a function, we pass it the icon so that it can be posisitioned whereever it is needed. If it is a string then the icon is automatically added next to it. If it is a component then a flex box wrap it automatically and the icon is centered left to that component.

JBWatenbergScality commented on src/react/account/AwsPaginatedResourceTable.tsx:
```diff
@@ -0,0 +1,186 @@
+                    singular: singularResourceName,
+                    plural: pluralResourceName,
+                  }}
+                />
```
I think we are still missing the disabled state here while loading the data as simiiar to:
`disabled={queryResult.status !== 'success'} `

JBWatenbergScality reviewed (commented):
A small comment remaining. Also the deletion test should likely be extended a bit to test the modal and the pensieve api call but maybe this can be done later on.

ChengYanJin commented:
> A small comment remaining. Also the deletion test should likely be extended a bit to test the modal and the pensieve api call but maybe this can be done later on.

I agree with this. I just created a ticket to track the missing test for policy deletion.

AsmaaELM commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.4.9`

* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.




*The following options are set:* **approve**

AsmaaELM commented:
ping

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ZKUI-161 contains:


* `1.5.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.4.9`

* `1.5.0`


Please check the `Fix Version/s` of ZKUI-161, or the target
branch of this pull request.




*The following options are set:* **approve**

ChengYanJin commented:
/status

bert-e commented:
# Status





*Status report is not available.*






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/1.5/feature/ZKUI-161-listing-policy-names` with contents from `feature/ZKUI-161-listing-policy-names`
and `development/1.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/1.5/feature/ZKUI-161-listing-policy-names origin/development/1.5
 $ git merge origin/feature/ZKUI-161-listing-policy-names
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/1.5/feature/ZKUI-161-listing-policy-names
```





*The following options are set:* **approve**

ChengYanJin commented:
/status

bert-e commented:
# Status





*Status report is not available.*






*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.4`

* :heavy_check_mark: `development/1.5`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.4`

* :heavy_check_mark: `development/1.5`




Please check the status of the associated issue ZKUI-161.

Goodbye asmaaelm.
